src/Swarm/Game/Step.hs

```diff
@@ -125,64 +132,62 @@ gameTick = do
 ------------------------------------------------------------
 
 -- | Perform operation on the monad with game state.
-doOnGame :: MonadState GameState m => m a -> ExceptT Exn (StateT Robot m) a
-doOnGame = lift . lift
 
 -- | Set a flag telling the UI that the world needs to be redrawn.
-flagRedraw :: MonadState GameState m => ExceptT Exn (StateT Robot m) ()
-flagRedraw = doOnGame $ needsRedraw .= True
+flagRedraw :: (Has (State GameState) sig m) => m ()
+flagRedraw = needsRedraw .= True
 
 -- | Perform an action requiring a 'W.World' state component in a
 --   larger context with a 'GameState'.
-zoomWorld :: MonadState GameState m => (forall n. MonadState (W.World Int Entity) n => n a) -> m a
+zoomWorld :: (Has (State GameState) sig m) => StateC (W.World Int Entity) Identity b -> m b
 zoomWorld n = do
   w <- use world
-  (a, w') <- runStateT n w
+  let (w', a) = run (runState w n)
   world .= w'
   return a
 
 -- | Get the entity (if any) at a given location.
-entityAt :: MonadState GameState m => V2 Int64 -> ExceptT Exn (StateT Robot m) (Maybe Entity)
-entityAt loc = doOnGame $ zoomWorld (W.lookupEntityM (W.locToCoords loc))
+entityAt :: (Has (State GameState) sig m) => V2 Int64 -> m (Maybe Entity)
+entityAt loc = zoomWorld (W.lookupEntityM @Int (W.locToCoords loc))
 
 -- | Modify the entity (if any) at a given location.
 updateEntityAt ::
-  MonadState GameState m =>
+  (Has (State GameState) sig m, Has (State Robot) sig m, Has (Throw Exn) sig m) =>
   V2 Int64 ->
   (Maybe Entity -> Maybe Entity) ->
-  ExceptT Exn (StateT Robot m) ()
-updateEntityAt loc upd = doOnGame $ zoomWorld (W.updateM (W.locToCoords loc) upd)
+  m ()
+updateEntityAt loc upd = zoomWorld (W.updateM @Int (W.locToCoords loc) upd)
 
 -- | Get the robot with a given name (if any).
-robotNamed :: MonadState GameState m => Text -> ExceptT Exn (StateT Robot m) (Maybe Robot)
-robotNamed nm = doOnGame $ use (robotMap . at nm)
+robotNamed :: (Has (State GameState) sig m) => Text -> m (Maybe Robot)
+robotNamed nm = use (robotMap . at nm)
 
 -- | Manhattan distance between world locations.
 manhattan :: V2 Int64 -> V2 Int64 -> Int64
 manhattan (V2 x1 y1) (V2 x2 y2) = abs (x1 - x2) + abs (y1 - y2)
 
 -- | Generate a uniformly random number using the random generator in
 --   the game state.
-uniform :: (MonadState GameState m, UniformRange a) => (a, a) -> ExceptT Exn (StateT Robot m) a
+uniform :: (Has (State GameState) sig m, UniformRange a) => (a, a) -> m a
```

Comment:
Eventually, it would also be nice to write this in terms of some kind of randomness effect, which we could then interpret in terms of `State GameState`.